### PR TITLE
validate the kube-dns is nil AND add ignore rules from coredns

### PR DESCRIPTION
**Pull request by yz271544**
exclude the services that is kube-dns or coredns

### CODE REVIEW

File: agent/pkg/proxy/iptables.go [hunk 1]
```diff
@@ -143,10 +143,19 @@ func (proxier *Proxier) createIgnoreRules() (ignoreRules []iptablesEnsureInfo, e
...
 	kubeDNS, err := proxier.kubeClient.CoreV1().Services("kube-system").Get(context.Background(), "kube-dns", metav1.GetOptions{})
-	if err != nil {
-		return
+	if kubeDNS != nil && err == nil {
+		klog.V(4).Infof("ignored kubeDNS: %s", kubeDNS.Name)
+		ignoreRules = append(ignoreRules, ignoreRuleByService(kubeDNS))
+	}
+
+	// coredns service
+	kubeDNSList, err := proxier.kubeClient.CoreV1().Services("kube-system").List(context.Background(), metav1.ListOptions{LabelSelector: "k8s-app=kube-dns"})
```
Comment (Poorunga) [inline agent/pkg/proxy/iptables.go:152]: Is the service name of coredns also kube-dns?

Comment (yz271544) [reply agent/pkg/proxy/iptables.go:152]: the coredns service:
```shell
NAME      TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                        AGE     LABELS
coredns   ClusterIP   10.10.0.3    <none>        53/UDP,53/TCP,9153/TCP         4d18h   addonmanager.kubernetes.io/mode=Reconcile,k8s-app=kube-dns,kubernetes.io/name=coredns
```

Comment (Poorunga) [reply agent/pkg/proxy/iptables.go:152]: Okay

Comment (Poorunga) [reply agent/pkg/proxy/iptables.go:152]: But will this cause us to add the same rule twice?

On my cluster:
```
[root@k8s-master ~]# kubectl get svc -n kube-system
NAME       TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   16d
```

kubeDNS will add the rule for the first time, and kubeDNSList will add the rule for the second time.

**Review by yz271544 — COMMENTED** [review]
Should we promise on a label, such as proxyby=edgemesh, to define which service traffic should be hijacked by edgemesh?

**Comment by Poorunga** [comment]
Currently, edgemesh hijacks all services on the cluster ip CIDR. So what I consider is to filter the services that edgemesh does not want to hijack.

Maybe we can add a label to the service that we don't want to hijack, and then process the label in the code. However, I think this(let users add labels) will make users feel a little troublesome.

File: agent/pkg/proxy/iptables.go (hunk 1 above)
Comment (Poorunga) [reply agent/pkg/proxy/iptables.go:152]: In order to keep only one iptables rule, can we delete kubeDNS?

Comment (yz271544) [reply agent/pkg/proxy/iptables.go:153]: It is indeed possible, so I created a map to deal with the problem that multiple retrievals may cause chain duplication.

**Comment by yz271544** [comment]
> Currently, edgemesh hijacks all services on the cluster ip CIDR. So what I consider is to filter the services that edgemesh does not want to hijack.
> 
> Maybe we can add a label to the service that we don't want to hijack, and then process the label in the code. However, I think this(let users add labels) will make users feel a little troublesome.

It seems that we can use the filtering feature of label in this way, but this requires adding an agreed label.
```go
	// Other services we want to ignore...
	otherIgnoreServiceList, err := proxier.kubeClient.CoreV1().Services("*").List(context.Background(), metav1.ListOptions{LabelSelector: "proxyby!=edgemesh"})
	if err == nil && otherIgnoreServiceList != nil && len(otherIgnoreServiceList.Items) > 0 {
		for _, item := range otherIgnoreServiceList.Items {
			klog.V(4).Infof("ignored not containing proxyby=edgemesh label service: %s", item.Name)
			ignoreRulesIptablesEnsureMap[strings.Join([]string{item.Namespace, item.Name}, ".")] = &item
		}
	}
```

File: agent/pkg/proxy/iptables.go (hunk 1 above)
Comment (Poorunga) [reply agent/pkg/proxy/iptables.go:153]: Perfect

**Comment by Poorunga** [comment]
> It seems that we can use the filtering feature of label in this way, but this requires adding an agreed label.

Right, there is no problem, we can do it.

However, if some users only use edgemesh at the edge node (without deploying kube-proxy), it means that we need to manually add a agreed label to each service.

**Comment by yz271544** [comment]
Can I add this code after this pr?

**Comment by Poorunga** [comment]
> Can I add this code after this pr?

Yes, but I suggest you raise an issue first, describing clearly why you want to do this. Maybe others need to understand the background and your intentions.

**Comment by yz271544** [comment]
Okay, let me test it in my own environment first!

**Comment by Poorunga** [comment]
Nice~

File: agent/pkg/proxy/iptables.go (hunk 1 above)
Comment (fisherxu) [reply agent/pkg/proxy/iptables.go:153]:
```suggestion
	kubeDNSList, err := proxier.kubeClient.CoreV1().Services(metav1.NamespaceSystem).List(context.Background(), metav1.ListOptions{LabelSelector: "k8s-app=kube-dns"})
```

**Pull request merged**